h2/src/main/org/h2/mvstore/tx/TransactionStore.java

```diff
@@ -62,17 +63,38 @@
 
     private final DataType dataType;
 
-    private final BitSet openTransactions = new BitSet();
+    final AtomicReference<VersionedBitSet> openTransactions = new AtomicReference<>(new VersionedBitSet());
+
+    final AtomicReference<BitSet> committingTransactions = new AtomicReference<>(new BitSet());
 
     private boolean init;
 
-    private int maxTransactionId = 0xffff;
+    /**
+     * Soft limit on the number of concurrently opened transactions.
+     * Not really needed but used by some test.
+     */
+    private int maxTransactionId = MAX_OPEN_TRANSACTIONS;
+
+    /**
+     * Array holding all open transaction objects.
+     * Position in array is "transaction id".
+     */
+    private final AtomicReferenceArray<Transaction> transactions = new AtomicReferenceArray<>(MAX_OPEN_TRANSACTIONS);
 
     /**
      * The next id of a temporary map.
      */
     private int nextTempMapId;
 
+
+    /**
+     * Hard limit on the number of concurrently opened transactions
+     */
+    // TODO: introduce constructor parameter instead of a static field, driven by URL parameter
+    private static final int MAX_OPEN_TRANSACTIONS = 0x100;
```

Comment:
Old value was 65,535, new is 256. This limit may be too restricted.

Only 4 bytes (with compressed pointers on 64-bit VMs) is needed for each element in `transactions`. So overhead is not very noticeable, may be this limit should be increased from 256 to some higher value.

Reply:
1024? I'd love to see H2 holding up with 1024 concurrent transactions!!

Reply:
There was no problem even with 65,535 transactions if they are not to huge before this changes. And with this changes if `MAX_OPEN_TRANSACTIONS` is increased to 65,535 all works as before.